scilpy/denoise/asym_filtering.py

```diff
@@ -1,387 +1,480 @@
+    is_legacy: bool
+        Whether the legacy SH basis definition should be used.
+    full_basis: bool
+        Whether the input SH basis is full or not.
+    sphere_str: str
+        Name of the DIPY sphere to use for SH to SF projection.
+    sigma_spatial: float or None
+        Standard deviation of spatial filter. Can be None to replace
+        by mean filter, in what case win_hwidth must be given.
+    sigma_align: float or None
+        Standard deviation of alignment filter. `None` disables
+        alignment filtering.
+    sigma_angle: float or None
+        Standard deviation of the angle filter. `None` disables
+        angle filtering.
+    rel_sigma_range: float or None
+        Standard deviation of the range filter, relative to the
+        range of SF amplitudes. `None` disables range filtering.
+    disable_spatial: bool, optional
+        Replace gaussian filter by a mean filter for spatial filter.
+        The value from `sigma_spatial` is still used for setting the
+        size of the filtering window.
+    win_hwidth: int, optional
+        Half-width of the filtering window. When None, the
+        filtering window half-width is given by (6*sigma_spatial + 1).
+    exclude_center: bool, optional
+        Assign a weight of 0 to the center voxel of the filter.
+    device_type: string, optional
+        Device on which the code should run. Choices are cpu or gpu.
+    use_opencl: bool, optional
+        Use OpenCL for software acceleration.
+    patch_size: int, optional
+        Patch size for OpenCL execution.
+
+    References
+    ----------
+    [1] Poirier and Descoteaux, 2024, "A Unified Filtering Method for
+        Estimating Asymmetric Orientation Distribution Functions",
+        Neuroimage, https://doi.org/10.1016/j.neuroimage.2024.120516
     """
-    if use_gpu and have_opencl:
-        return angle_aware_bilateral_filtering_gpu(in_sh, sh_order,
-                                                   sh_basis, in_full_basis,
-                                                   is_legacy,
-                                                   sphere_str, sigma_spatial,
-                                                   sigma_angular, sigma_range)
-    elif use_gpu and not have_opencl:
-        raise RuntimeError('Package pyopencl not found. Install pyopencl'
-                           ' or set use_gpu to False.')
+    if sigma_spatial is None and win_hwidth is None:
+        raise ValueError('sigma_spatial and win_hwidth cannot both be None')
+    if device_type not in ['cpu', 'gpu']:
+        raise ValueError('Invalid device type {}. Must be cpu or gpu'
+                         .format(device_type))
+    if use_opencl and not have_opencl:
+        raise ValueError('pyopencl is not installed. Please install before'
+                         ' using option use_opencl=True.')
+    if device_type == 'gpu' and not use_opencl:
+        raise ValueError('Option use_opencl must be enabled '
```

Comment:
About your question on how to manage this. Another option could be, here, to print a warning that you force the change to --use_opencl because user chose gpu, but not raise an error.

Both options are ok for me.

Reply:
I'll keep it this way here and make the changes in the argparser